test/parse_test.js

```diff
@@ -377,6 +379,19 @@ describe('parse()', function() {
       await parser.parse(() => {});
     }, type, message);
   });
+
+  it('should properly mark circular references', async function() {
+    const result = await parser.parse(inputYAMLCircular, { path: __filename });
+    expect(result.components().schema('RecursiveAncestor').properties().children._json.items[xParserCircle]).to.equal(true);
+    expect(result.components().schema('RecursiveSelf').properties().something._json.properties.test.properties.children.items[xParserCircle]).to.equal(true);
+    expect(result.channel('external/file').publish().json().message.payload.properties.test.properties.children.items[xParserCircle]).to.equal(true);
+    expect(result.json()[xParserCircle]).to.equal(true);
+    //not testing on a model level as required xParserCircle value is added before model construction so we need to test through calling parser function
+    expect(result.hasCircular()).to.equal(true);
+    expect(result.components().schema('RecursiveSelf').isCircular()).to.equal(true);
```

Comment:
Ok so now I see where we have different understandings of the meaning of "circular". IMO, this should evaluate to `false` and instead `hasCircular()` should evaluate to `true`. Why?

1. This schema is not circular, it's recursive. It **contains** circular properties. More specifically, the `children` property **is** circular. Even more, the `items` schema **is** circular.
1. As a developer, I'd love to know when I'm in front of a recursive schema and when I'm in front of a circular property of that schema. As an example, in the HTML renderer, I want to render the schema as usual, with all its properties but once I get into the circular property, I don't want to continue rendering nested properties but instead, I want to just render a link to another part of the document (where the referenced schema is rendered). The same would be applied to the generation of types/structs/etc. If we don't do it like this, developers would have to check themselves if that's the first time they see this schema or not.

Following what I said, tests should look like:

```js
    expect(result.components().schema('RecursiveSelf').isCircular()).to.equal(false);
    expect(result.components().schema('RecursiveSelf').hasCircular()).to.equal(true);
    expect(result.components().schema('RecursiveSelf').properties()['children'].items().isCircular()).to.equal(true);
```

Reply:
```
    expect(result.components().schema('RecursiveSelf').isCircular()).to.equal(false);
    expect(result.components().schema('RecursiveSelf').hasCircular()).to.equal(true);
```

makes total sense, totally agree as what you wrote, it has, not is

Reply:
@fmvilas do you think `expect(result.components().schema('RecursiveSelf').hasCircular()).to.equal(true);` is a must? like useful really. I can see a use case for `expect(result.hasCircular()).to.equal(true);` as the tool can quickly throw to the user that circular refs are not supported, but `expect(result.components().schema('RecursiveSelf').hasCircular()).to.equal(true);` doesn't help much as it means that some properties might be circular and some not and I anyway have to go throught them and react only in case of `isCircular()`

or you provided a use case above but I just didn't fully get it?

Reply:
That's a good point. I can't see it being used anywhere. You can probably drop `hasCircular` from Schema model.

Reply:
@fmvilas ok man, it's completed. Works like a charm.

I want to admit, that after looking at the implementation now I agree now that this `isCircular` function can be very helpful for tools developers!